### ISSUE TITLE
GitHub integration 

### ISSUE THREAD

**Issue by bellaabdelouahab** [ISIL-ESTE/Student-Workflow-Organizer#58]
This issue aims to implement the initial steps for GitHub integration and retrieve repositories
this will unlock many possibilities in our application.
here are tasks to be done :
- [ ] Implement authentication mechanisms to securely connect to the GitHub API.
- [ ] Store the required access token securely for future API requests
- [ ] Implement a mechanism to retrieve the user's recent repository from their GitHub account, and extract essential information such as repository name, description
- [ ] this is an extra, and yet an important task signup to the application via GitHub.

**Comment by bellaabdelouahab** [comment]
@muttaqin1 do you want to take this ?

**Comment by muttaqin1** [comment]
I havent done github integration before. This will be a big challenge for me.But i will try.
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
sure, but it's not that hard the hard part is managing access to certain things like email username repositories and other info.
@muttaqin1

**Comment by muttaqin1** [comment]
I have to use passport.js right?
Then when a user log's in i will get a access token and refresh token i will store that in db so that i can do some api request's to Github api?

**Comment by muttaqin1** [comment]
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
here are some resources 
https://codevoweb.com/github-oauth-authentication-react-and-node/
@muttaqin1

**Comment by muttaqin1** [comment]
yeah, I have read the blog. It was beneficial thank you. 
@bellaabdelouahab

**Comment by muttaqin1** [comment]
Hello @bellaabdelouahab, 
Github Oauth doesn't give us access to user password. In that case i am willing to create a model to store github user info and access token.
But if we allow null value on the password property in our USER model we can store the data which we will get from github.Then we have to create another model to store access token.
So i believe the first option will be great for our use case.What do you think?

**Comment by bellaabdelouahab** [comment]
Hey @muttaqin1,
Yes of course what we need to do is either two ways allow access without password or after connecting to github require the user to enter his password both options are good.

**Comment by bellaabdelouahab** [comment]
You don't need to create a new model just add the field of github access token to existing userModel

**Comment by bellaabdelouahab** [comment]
Another thing,
Don't store data like repositories or projects...
Just the needed data to create a user,
About the repos and other info we will make it stateless.

**Comment by muttaqin1** [comment]
> Another thing,
> Don't store data like repositories or projects...
> Just the needed data to create a user,
> About the repos and other info we will make it stateless. 

Yeah i know that.

**Comment by bellaabdelouahab** [comment]
@muttaqin1  any heads-up ?

**Comment by muttaqin1** [comment]
> @muttaqin1  any heads-up ? 

Yeah, 90% done and now i'm fixing some bugs.I will submit a pr soon.
@bellaabdelouahab

**Comment by muttaqin1** [comment]
I have generated the jwt  and set it to the browser using a secured cookie.
then redirected to the path which i have got from the req.params.

```js
const {code, path} = req.params

//...some tasks
//... generate jwt
//...set cookie

res.redirect(path);
```
Is there any other way where we don't have to use the cookie?

like this:
```js
const {code} = req.params

//...some tasks
//... generate jwt

res.status(200).json({
user:createdUser,
token
})
```
But whenever the callback url is called by github. The server response open's in the browser. Because its a GET request.


@bellaabdelouahab

**Comment by muttaqin1** [comment]
![Screenshot_2023-07-12-18-38-52-47.jpg](https://github.com/ISIL-ESTE/Student-Workflow-Organizer/assets/82669867/6feb8128-caca-45b2-9292-c3c04e11a06c)

@bellaabdelouahab
